File: Pymodel/NTT_Kara_16/NTT8Kara8_final/np_version/Poly_NTT8_Kara8.py
import time
import numpy as np
from NTT8_Kara8_helper import *

# Parameter Definition
N       =  65536
K       =  32
q       =  998244353
phi     =  629671588
phi_inv =  283043518
psi     =  24514907
psi_inv =  3707709
N_inv   =  998229121
inv_2   =  499122177

np.set_printoptions(linewidth=168)

start_time = time.time()
print("start_time: ", start_time)

# -----------------------------------
print("********** Start Init **********")
A = np.arange(N, dtype=np.int64)
B = np.arange(N, dtype=np.int64)
# A = np.random.randint(0, q, size=N, dtype=np.int64)
# B = np.random.randint(0, q, size=N, dtype=np.int64)
print("A[0:16]: ")
print(A[0:16])
print("B[0:16]: ")
print(B[0:16])
print("********** End Init **********")
print("\n\r")

# -----------------------------------
print("********** Start Nega_conv **********")
nega_conv_start_time = time.time()

indices = np.arange(N, dtype=np.int64)
# psi_powers = np.power(psi, indices, dtype=np.int64) % q
psi_powers = np.array([pow(psi, int(i), q) for i in indices], dtype=np.int64)

A_nega = (A * psi_powers) % q
B_nega = (B * psi_powers) % q
print("A_nega[0:16]: ")
print(A_nega[0:16])
print("B_nega[0:16]: ")
print(B_nega[0:16])

nega_conv_end_time = time.time()
print("Nega_conv Time: ", nega_conv_end_time - nega_conv_start_time)
print("********** End Nega_conv **********")
print("\n\r")


# -----------------------------------
print("********** Start NTT **********")
ntt_start_time = time.time()

A_ntt8 = NTT8(A_nega, q, phi)
B_ntt8 = NTT8(B_nega, q, phi)
print("A_ntt8[0:16]: ")
print(A_ntt8[0:16])
print("B_ntt8[0:16]: ")
print(B_ntt8[0:16])

ntt_end_time = time.time()
print("NTT Time: ", ntt_end_time - ntt_start_time)
print("********** End NTT **********")
print("\n\r")
# #### WRONG CMODEL ####
print("A_ntt8[240:256]: ")
print(A_ntt8[240:256])
print("A_ntt8[256:272]: ")
print(A_ntt8[256:272])

# -----------------------------------
print("********** Start Kara_Poly Kernel **********")
kara_start_time = time.time()

A_kara8   = np.zeros(6561, dtype=np.int64)
B_kara8   = np.zeros(6561, dtype=np.int64)
C_mult    = np.zeros(6561, dtype=np.int64)
C_block   = np.zeros(256, dtype=np.int64)
C_ikara8  = np.zeros(N, dtype=np.int64)

for i in range(256):
    A_block = A_ntt8[i*256:(i+1)*256]
    B_block = B_ntt8[i*256:(i+1)*256]
    
    # Karatsuba
    A_kara8 = Kara8(A_block, q)
    B_kara8 = Kara8(B_block, q)
    
    # Element-wise Multiplication
    C_mult = np.multiply(A_kara8, B_kara8) % q
    
    # Inverse Karatsuba
    C_block = Inv_Kara8(C_mult, q)
    
    C_ikara8[i*256:(i+1)*256] = C_block

print("C_ikara8[0:16]: ")
print(C_ikara8[0:16])

kara_end_time = time.time()
print("Karatsuba Time: ", kara_end_time - kara_start_time)
print("********** End Kara_Poly Kernel **********")
print("\n\r")

# -----------------------------------
print("********** Start Inv_NTT **********")
intt_start_time = time.time()

C_intt8 = Inv_NTT8(C_ikara8, q, phi_inv, inv_2)
print("C_intt8[0:16]: ")
print(C_intt8[0:16])

intt_end_time = time.time()
print("INTT Time: ", intt_end_time - intt_start_time)
print("********** End Inv_NTT **********")
print("\n\r")

# -----------------------------------
print("********** Start Inv_Nega_conv **********")
inv_nega_conv_start_time = time.time()

indices = np.arange(N, dtype=np.int64)
# psi_inv_powers = np.power(psi_inv, indices, dtype=np.int64) % q
psi_inv_powers = np.array([pow(psi_inv, int(i), q) for i in indices], dtype=np.int64)

C = (C_intt8 * psi_inv_powers) % q
print("C[0:16]: ")
print(C[0:16])

inv_nega_conv_end_time = time.time()
print("Inv_Nega_conv Time: ", inv_nega_conv_end_time - inv_nega_conv_start_time)
print("********** End Inv_Nega_conv **********")
print("\n\r")

# -----------------------------------
print("********** Start Verification **********")

Base_temp = np.zeros(2*N-1, dtype=np.int64)
for i in range(N):
    Base_temp[i:i+N] += (A[i]*B) % q
    # Base_temp[i+j] %= q
Base_temp %= q
Base = np.zeros(N, dtype=np.int64)
Base[:N] = Base_temp[:N]
Base[:N-1] -= Base_temp[N:]
Base %= q
print("Base[0:16]: ")
print(Base[0:16])

if np.array_equal(C, Base):
    print("Verification PASS! :) :) :) :) :)")
else:
    print("Verification FAIL! :( :( :( :( :(")
print("********** End Verification **********")
print("\n\r")

end_time = time.time()
print("end_time: ", end_time)
print("Execution Time: ", end_time - start_time)

File: Cmodel/N8K8/Poly_NTT8_Kara8.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <inttypes.h>
#include <assert.h>

#include "NTT8_Kara8_helper.h"

// Parameter Definition
#define N            65536       // N = 2^16
// #define N_kara       1679616     // N_kara = 2^8 * 3^8
#define Kara_group   256
#define Kara_256     256
#define Kara_6561    6561
#define K            32
#define q            998244353
#define phi          629671588
#define phi_inv      283043518
#define psi          24514907
#define psi_inv      3707709
#define N_inv        998229121
#define inv_2        499122177

/*
 * Print array of size n
 */
static void print_array(FILE *f, int32_t *a, int32_t n) {
  uint32_t i, k;

  k = 0;
  for (i=0; i<n; i++) {
    if (k == 0) fprintf(f, "  ");
    fprintf(f, "%5"PRId32, a[i]);
    k ++;
    if (k == 16) {
      fprintf(f, "\n");
      k = 0;
    } else {
      fprintf(f, " ");
    }
  }
  if (k > 0) {
    fprintf(f, "\n");
  }
}

static int32_t random_coeff(void) {
  int32_t x;

  x = rand() % q;
  assert(0 <= x && x <= q-1);
  return x;
}

int main() {
    // -----------------------------------------
	  printf("********** Start Init **********\n");
    uint32_t i,j;
	  uint32_t A[N] = {0};
    uint32_t B[N] = {0};
    uint32_t C[N] = {0};
    // for (i = 0; i < N; i++) {
    //     A[i] = i;
    //     B[i] = i;
    // }
    for (i = 0; i < N; i++) {
        A[i] = random_coeff();
        B[i] = random_coeff();
    }
    printf("A[0:15]");
    print_array(stdout, A, 16);
    printf("B[0:15]");
    print_array(stdout, B, 16);
    printf("********** End Init **********\n\n");

    // -----------------------------------------
    printf("********** Start Nega_conv **********\n");
    Nega_conv(A, N, psi, q);
    Nega_conv(B, N, psi, q);
    printf("A[0:15]");
    print_array(stdout, A, 16);
    printf("B[0:15]");
    print_array(stdout, B, 16);
    printf("********** End Nega_conv **********\n\n");

    // -----------------------------------------
    printf("********** Start NTT **********\n");
    NTT_8(A, N, phi, q);
    NTT_8(B, N, phi, q);
    printf("A[0:15]");
    print_array(stdout, A, 16);
    printf("B[0:15]");
    print_array(stdout, B, 16);
    printf("********** End NTT **********\n\n");

    // -----------------------------------------
    printf("********** Start Karatsuba - Element-wise Mult **********\n");
    uint32_t A_6561[Kara_6561] = {0};
    uint32_t B_6561[Kara_6561] = {0}; 
    uint32_t C_6561[Kara_6561] = {0}; 

    for (i = 0; i < Kara_group; i++) {
        uint32_t* A_256  = &A[i*Kara_256];
        uint32_t* B_256  = &B[i*Kara_256];
        Kara_8(A_256, A_6561, q);
        Kara_8(B_256, B_6561, q);
        // -----------------------------------------
        Ele_wise_Mult(A_6561, B_6561, C_6561, Kara_6561, q);
        // -----------------------------------------
        uint32_t* C_256  = &C[i*Kara_256];
        Inv_Kara_8(C_256, C_6561, q);
    }
    printf("C[0:15]");
    print_array(stdout, C, 16);
    printf("********** End Karatsuba - Element-wise Mult **********\n\n");

    // -----------------------------------------
    printf("********** Start Inv_NTT **********\n");
    Inv_NTT_8(C, N, phi_inv, inv_2, q);
    printf("C[0:15]");
    print_array(stdout, C, 16);
    printf("********** End Inv_NTT **********\n\n");

    // -----------------------------------------
    printf("********** Start Inv_Nega_conv **********\n");
    Nega_conv(C, N, psi_inv, q);
    printf("C[0:15]");
    print_array(stdout, C, 16);
    printf("********** End Inv_Nega_conv **********\n\n");

    // -----------------------------------------
    printf("********** Start Verification **********\n");
    uint32_t Base_temp[2*N-1] = {0};
    uint32_t Base[N] = {0};
    for (i = 0; i < N; i++) {
      for (j = 0; j < N; j++) {
        Base_temp[i+j] = Base_temp[i+j] + (uint64_t)A[i] * B[j] % q;
      }
    }
    for (i = 0; i < N; i++) {
      Base[i] = Base_temp[i] % q;
    }
    for (i = N; i < 2*N-1; i++) {
      Base[i-N] = (Base[i-N]+q - Base_temp[i]) % q;
    }

    if (memcmp(C, Base, N * sizeof(uint32_t)) == 0) {
      printf("Verification PASS! :) :) :) :) :)\n");
    } else {
      printf("Verification PASS! :) :) :) :) :)\n");
    }

    printf("********** End Verification **********\n\n");     
      
    return 0;
}
